8357086: os::xxx functions returning memory size should return size_t

toxaart opened the pull request.

stefank commented on src/hotspot/os/linux/cgroupSubsystem_linux.cpp:
```diff
@@ -669,9 +669,9 @@ jlong CgroupSubsystem::memory_limit_in_bytes() {
-  log_trace(os, container)("total physical memory: " JLONG_FORMAT, phys_mem);
-  jlong mem_limit = contrl->controller()->read_memory_limit_in_bytes(phys_mem);
+  size_t phys_mem = os::Linux::physical_memory();
+  log_trace(os, container)("total physical memory: " "%zu", phys_mem);
```
```suggestion
  log_trace(os, container)("total physical memory: %zu", phys_mem);
```

toxaart replied:
Addressed in the latest commit.

stefank commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -259,52 +259,52 @@ julong os::Linux::available_memory() {
 }
 
-static jlong host_free_swap() {
+static size_t host_free_swap() {
```
It is not obvious to me that callers of this are reacting properly when the return value goes from signed to unsigned. I'm thinking about code in os::free_swap_space below:
```
jlong host_free_swap_val = MIN2(os::total_swap_space(), host_free_swap());
```
How does MIN2 react to -1 vs 0xFFFFFFFFFFFFFFFF

stefank commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -332,21 +332,21 @@ jlong os::free_swap_space() {
 
   phys_mem = Linux::physical_memory();
-  log_trace(os)("total system memory: " JLONG_FORMAT, phys_mem);
+  log_trace(os)("total system memory: " "%zu", phys_mem);
```
```suggestion
  log_trace(os)("total system memory: %zu", phys_mem);
```

stefank commented on src/hotspot/share/gc/z/zLargePages.cpp:
```diff
@@ -31,7 +31,7 @@ bool ZLargePages::_os_enforced_transparent_mode;
   pd_initialize();
 
-  log_info_p(gc, init)("Memory: " JULONG_FORMAT "M", os::physical_memory() / M);
+  log_info_p(gc, init)("Memory: %zu" "M", os::physical_memory() / M);
```
```suggestion
  log_info_p(gc, init)("Memory: %zuM", os::physical_memory() / M);
```

stefank reviewed (commented):
Thanks for fixing the signed vs unsigned types, I'm much more confident with that.

I talked to Anton offline about the `ptrdiff_t`. That type has the correct signedness and number of bits on all our platforms, but to me that type carries a semantic meaning about what pointer diffs / indices. Because of that I found it inappropriate to use that type. I was expecting to see `ssize_t` being used, but that type is only guaranteed to hold [-1, SSIZE_MAX] and we have paths through the container code that returns -2. The man pages say "It is a signed integer type capable of storing values at least in the range [-1, SSIZE_MAX]", so there's an "at least" which might indicate that some systems allow -2? Maybe all our platform does that? Could that be tested with some kind of static asserts or unit test? Could others comment on what the correct type should be for these functions? Maybe @kimbarrett has a good idea about what type to use?

An alternative could be to let these functions return a compound value [size_t size, int error_code], or some adhoc optional type. I think that would be the safest way forward for these.

Looking forward to here from other reviewers.

Anton, I know that you intentionally left the code to be `"%zu" "k"` and I'm not going to argue strongly against that in the RT code, but I don't really want that in the ZGC code (or the GC code) so I added a couple of suggestions to change that.

dholmes-ora commented:
> I talked to Anton offline about the ptrdiff_t. That type has the correct signedness and number of bits on all our platforms, but to me that type carries a semantic meaning about what pointer diffs / indices. Because of that I found it inappropriate to use that type. 

Those were my thoughts as well when I saw that change made. I don't have a good answer either. It is not at all clear how/why the swap functions were allowed to report errors when none of the others do. It is not documented in os.hpp nor required by the `OperatingSystemMXBean` API!

> we have paths through the container code that returns -2

I couldn't spot these - do you have a link please? It is annoying this prevents use of `ssize_t`.

> An alternative could be to let these functions return a compound value [size_t size, int error_code],

Or, as is done in other code, use an out-parameter e.g:
```
// Returns the amount of free swap. If an error occurs then `rc` is set to a negative value
// and we return zero.; otherwise `rc` is set to zero.
static size_t free_swap_space(int& rc);
```
?

dholmes-ora commented:
> It is not documented in os.hpp nor required by the OperatingSystemMXBean API!

So I just discovered the swap functions (possibly others) are not actually used by the `OperatingSystemMXBean` API. That code exists on the JDK side.

toxaart commented:
> I couldn't spot these - do you have a link please? It is annoying this prevents use of `ssize_t`.

In osContainer_linux.hpp there is `OSCONTAINER_ERROR` defined with value -2. This may be returned, for instance, in `OSContainer::memory_and_swap_limit_in_bytes() `, which unrolls to `jlong CgroupV1MemoryController::memory_and_swap_limit_in_bytes(julong host_mem, julong host_swap)`, and there is a call to `read_memory_limit_in_bytes(host_mem)`, and that method can return the error code.

toxaart commented:
Another thing to keep in mind: some methods in container-related code use -1 for unlimited value, -2 for error.

toxaart commented:
@tstuefe do you have any suggestions?

stefank reviewed (changes requested):
Are you up for making an experiment that changes `total_swap_space` and `free_swap_space` to return two values: one the actual value in `size_t` and  the other an error code that gets set whenever we hit an error?

The two proposed alternatives for this would be:
1) returning something containing the two values (struct, Pair, Tuple, array, ...)
2) Return the size_t and using an an out put parameter to signal an error (or vice versa)

I think the fan-out of that will not be too bad and the likely outcome is that it is clearer that the code is propagating errors.

kimbarrett commented:
Windows doesn't seem to have ssize_t natively, so we define it and SSIZE_MAX and SSIZE_MIN ourselves
(in globalDefinitions_visCPP.hpp), with the obvious definitions.

jerboaa reviewed (commented):
+1 to the `ssize_t` change. Much better than the `ptrdiff_t` version.

toxaart commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -258,55 +258,55 @@ julong os::Linux::available_memory() {
   }
   log_trace(os)("available memory: " JULONG_FORMAT, avail_mem);
-  return avail_mem;
+  return static_cast<size_t>(avail_mem);
```
Thanks for spotting this. Line 243 definitely needs treatment as well as line 258. Addressed in the latest commit. 

Line 257 does not look redundant to me. `avail_mem` could be set to large value (cast from -1) on line 240, then lines between 246 and 256 can actually left untouched if something went wrong with reading meminfo file. Then we need to check if we still have an error value in `avail_mem`, and if so, use the one returned by `free_memory()`.

stefank replied:
I agree. (Though I still think this method could be improved to just use one -1 check. Obviously not in this PR).

kimbarrett commented:
> > D) a separate error reporting variable. It feels a bit overengineered, though.
> 
> I think I disagree with this. The status quo makes it all too easy to miss that you need to check the return value for an error. IMHO, this is a growing ground for bugs when someone starts to use any of these APIs with realizing that we're mixing two values (and two types) into one return value. Just 2c, to reiterate that _I_ wouldn't mind a solution like this.

I agree with @stefank here.  Also, these are not functions that would be helped by C++17 `[[nodiscard]]`.

Note that we already have a mechanism for propagating errors in HotSpot, via CHECK/TRAPS. Not beautiful but it
exists. This discussion makes me kind of wish we had something like [Boost.LEAF](https://boostorg.github.io/leaf/).
(Though probably trimmed down a bit, since there are features we don't need. And having anything like that likely
needed a more modern C++ than we had until relatively recently.)

toxaart commented:
> I prefer the dedicated struct approach, as it
> provides meaningful names.

I now refactored the code to return a struct instead. 

> This may have been more effort than I envisioned when creating this issue, sorry for that. I would love it if we could unify these behaviours.

I tried to address all the issues you pointed out. Now all methods return the same type, and on all platforms including Windows there is a check.

toxaart commented:
> FWIW, I think that the `ssize_t` was a good first step and the `MemRes` was an experiment that would be interesting to see if how that panned out, but I didn't expect to see it in this PR. It tried to convey that in an earlier comment. I'm leaving it up to the other involved reviewers to decide if we should go with the MemRes change in this PR.

I understand that. Keeping `ssize_t` would not introduce consistency in the codebase, as on Windows `ssize_t` is `int64_t`, which is a signed type and, as pointed out by @tstuefe, it eats half of the possible range. I am not sure about other platforms, but I suspect that would not be the case there. Therefore I decided to rely on the known things such as `size_t `. But then, if we want to report more than one kind of error, it makes sense to keep the error code separately. 

> Tons of MemErr's without examining the error part at all. I guess that's a
> consequence of the existing code just not bothering to check, which I guess
> needs to be fixed but not in this PR.

I agree, checking of error codes and making sure it is not easy to ignore them is to be done in a separate PR.

toxaart commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -145,6 +145,14 @@ enum WXMode {
+// struct to return from mem-related functions
+struct MemRes {
+  size_t val;
+  int err;
```
I changed the variables names to full words for better readability, thanks.

toxaart commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -145,6 +145,14 @@ enum WXMode {
+  size_t val;
+  int err;
+  MemRes(size_t v, int e) : val(v), err(e) {}
+  MemRes(): val(0), err(0) {}
```
I removed the default constructor, as decided to construct in the return statement with given argument(s).

toxaart commented:
> About the MemRes change. I think adding the MemRes instance and then initializing it later makes the code messier:

I generally agree that it is more convenient to construct in the return statement. However, to be consistent with the rest of the codebase, I decided to use a call the constructor (not just the braces version, though it also works).

tstuefe commented:
> > FWIW, I think that the `ssize_t` was a good first step and the `MemRes` was an experiment that would be interesting to see if how that panned out, but I didn't expect to see it in this PR. It tried to convey that in an earlier comment. I'm leaving it up to the other involved reviewers to decide if we should go with the MemRes change in this PR.
> 
> I understand that. Keeping `ssize_t` would not introduce consistency in the codebase, as on Windows `ssize_t` is `int64_t`, which is a signed type and, as pointed out by @tstuefe, it eats half of the possible range. I am not sure about other platforms, but I suspect that would not be the case there. Therefore I decided to rely on the known things such as `size_t `. But then, if we want to report more than one kind of error, it makes sense to keep the error code separately.

What caller really needs multiple error codes? We want to know a memory size, and we may get it or not. If we don't get it, I don't think it matters why we don't get it. I don't see any different actions taken *by the caller of os::xxx()* in reaction to what cog in the machine exactly failed. The end user needs a way to get detail information, but that can be done with UL logging down in the function itself. In addition, for very severe errors the function could end the JVM right away themselves, no need to even return to the caller. Example: if /proc is missing. 

If we don't end the JVM and return false to the caller, all he then needs to decide is if the missing value is important to him. Printing functions would not care beyond writing "no data" or some such. Java heap initialization may want to terminate the JVM early if we cannot e.g. determine the physical_memory() size.

dholmes-ora commented:
> Note that we already have a mechanism for propagating errors in HotSpot, via CHECK/TRAPS. 

@kimbarrett  That is not a general error reporting mechanism, it is only for checking for Java exceptions.

kimbarrett commented:
> > Note that we already have a mechanism for propagating errors in HotSpot, via CHECK/TRAPS.
> 
> @kimbarrett That is not a general error reporting mechanism, it is only for checking for Java exceptions.

Oh, you are right, I forgot about that. It seems I hardly ever deal with that mechanism, and not at all recently.

toxaart commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -145,6 +145,13 @@ enum WXMode {
   WXExec
 };
 
+// struct to return from mem-related functions
```
> Can you add some more commentary on this please: what do the fields mean? how are they used? Do we need an assert to check that there is no error if you read the value?

The `value` field transfers a value returned by a call to memory functions (or computed by some other logic, see for instance the container case). The `error` field is introduced in order to signal an error which happened during a call to a memory function. Currently, the presence of error is encoded as -1 in that field, and the absence of error is 0. One could also do it with a boolean value. If later one wants to distinguish between different types of error, more values can be used. 

Having these two values separately allows to have the full range of `size_t` in the output of all functions of interest, and be able to signal if something went wrong at the same time. 

> It seems most places are not even checking if there may be an error. ??

This is correct, the purpose of this PR is not to enhance/improve handling of errors, but to address the issue of memory functions having different j* types where C++ doesn't interface with Java code.

toxaart commented:
Okay, let me summarize the findings: 

1) One needs to be consistent with respect to the type of the return value, of `os::xxx()` and having `size_t` in all methods is something everyone agrees on. 
2) There is a consensus that errors should be reported and handled properly, not ignored. 
3) There is only one type of error.
4) The usage pattern should make it difficult for the user to ignore the error if it is reported. 

Given the above, I think the optimal solution is to have a boolean as a return type to indicate the error (true for no error, false for error), and the actual memory value as an in-parameter transferred by reference or pointer. The usage pattern may be enforced by `nodiscard` attribute, but it available from C++17 only. For now, one can just add if/else statements, and add he attribute later after upgrade to C++17.

toxaart commented:
/touch

kimbarrett reviewed (commented):
Maybe we should have an `ATTRIBUTE_NODISCARD` macro with an always empty
definition, and add it to all the functions being changed? Otherwise, once
`[[nodiscard]]` becomes available, someone is going to have to hunt down all
the relevant functions. Once we have C++17 we can revisit. It might take some
preliminary call-site fixes before we can change the macro (or more likely,
remove it and just use `[[nodiscard]]` directly).

@toxaart - Before you do anything about that, what do other reviewers think?

stefank commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -255,98 +256,122 @@ julong os::Linux::available_memory() {
   if (OSContainer::is_containerized()) {
     if (OSContainer::memory_limit_in_bytes() > 0) {
-      return (jlong)(OSContainer::memory_and_swap_limit_in_bytes() - OSContainer::memory_limit_in_bytes());
+      value = static_cast<size_t>(OSContainer::memory_and_swap_limit_in_bytes() - OSContainer::memory_limit_in_bytes());
```
I might have wondered about this before, but can't `memory_and_swap_limit_in_bytes` return `-1` even when `memory_limit_in_bytes > 0` since -1 is treated as an "unlimited limit" here:
```
jlong CgroupV1MemoryController::read_mem_swap(julong host_total_memsw) {
  julong memswlimit;
  CONTAINER_READ_NUMBER_CHECKED(reader(), "/memory.memsw.limit_in_bytes", "Memory and Swap Limit", memswlimit);
  if (memswlimit >= host_total_memsw) {
    log_trace(os, container)("Memory and Swap Limit is: Unlimited");
    return (jlong)-1;
  }
```
Maybe this will be handled in a cleanup to the container subsystem?

toxaart replied:
I think we have discussed it above. In `os::free_swap_space(size_t& value)` implementation in `os_linux.cpp` the value returned by `OSContainer::memory_and_swap_limit_in_bytes()` is kept in a signed variable of `jlong` type. Then, if that value is negative, `host_free_swap_val` value will be returned, as that one is non-negative.

stefank replied:
My comment above is for the `total_swap_space` function and not the `free_swap_space` function.

stefank commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -1465,11 +1471,19 @@ void os::print_memory_info(outputStream* st) {
-  st->print("(" UINT64_FORMAT "k free)",
-            os::available_memory() >> 10);
+  size_t phys_mem = 0;
+  if (!os::physical_memory(phys_mem)) {
+    log_debug(os, thread)("os::physical_memory() failed");
+  }
```
FWIW, I find all these UL logging statements in the os::xxx_memory functions highly dubious, and think that it would be better to have some carefully thought-through places where we log system memory information. The spread-out of the UL logging in the various memory functions look like debugging aid for HotSpot devs and not really for the users of the JVM. So, maybe they could be at most be develop logging statements. I think this could be a discussion that needs to be done outside of this PR, but since even more logging statements are added with this PR I'm bringing this up here. Could we at least skip adding these log lines in this PR? The same comment goes for the rest of the platform files.

toxaart replied:
Logging removed, a proper one to be added back later.

stefank commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -848,40 +848,61 @@ jlong os::elapsed_frequency() {
-  return win32::physical_memory();
+bool os::physical_memory(size_t& value) {
+  size_t phys_mem = win32::physical_memory();
+  if (phys_mem == std::numeric_limits<size_t>::max())
+  {
```
```suggestion
  if (phys_mem == std::numeric_limits<size_t>::max()) {
```

stefank commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -281,7 +287,7 @@ void os::Bsd::initialize_system_info() {
   if (sysctl(mib, 2, &mem_val, &len, nullptr, 0) != -1) {
     assert(len == sizeof(mem_val), "unexpected data size");
-    _physical_memory = mem_val;
+    _physical_memory = static_cast<size_t>(mem_val);
```
It is also "interesting" that the BSD port doesn't propagate the failure to set up `_physical_memory` but the other platforms do. This adds an inconsistency across our platforms and I think that would be good to unify the behavior around that.

toxaart replied:
Thanks for spotting the missed code for __OpenBSD__ section. Addressed. 

I agree that BSD behaves somewhat differently compared to other platforms from this perspective. I suggest to create a separate issue and harmonize the behavior later separately.

stefank commented on src/hotspot/share/compiler/compileBroker.cpp:
```diff
@@ -1058,7 +1058,8 @@ void CompileBroker::possibly_add_compiler_threads(JavaThread* THREAD) {
   // Now, we do the more expensive operations.
-  julong free_memory = os::free_memory();
+  size_t free_memory = 0;
+  os::free_memory(free_memory);
```
I wonder if this should be changed to something like this:
```suggestion
  // Return value ignored - defaulting to 0 on failure.
  (void)os::free_memory(free_memory);
```
to clearly indicate the intent to skip the error check?

toxaart replied:
This is a valid point, but then we'll need to address it somehow else once a pattern with `[[nodiscard]]` or enum class return value is introduced. Or, simply add an empty clause for failure. For now, I will just put a comment there as suggested.

stefank replied:
Note that the main point here was the cast to (void) used to indicate that the return value is explicitly discarded. Does that not work with the [[nodiscard]]?

toxaart replied:
Sorry, I completely missed that point. Need to check this.

kimbarrett replied:
Explicitly casting to void is the canonical idiom for ignoring [[nodiscard]].

stefank commented on src/hotspot/share/gc/shared/gcInitLogger.cpp:
```diff
@@ -62,7 +62,12 @@ void GCInitLogger::print_cpu() {
 
 void GCInitLogger::print_memory() {
-  julong memory = os::physical_memory();
+  size_t phys_mem = 0;
+  if (!os::physical_memory(phys_mem))
+  {
+    log_debug_p(gc, init)("os::physical_memory() failed");
+  }
+  julong memory = static_cast<julong>(phys_mem);
```
I would prefer to not have the extra logging in the GC code + some extra restructuring:
```
  size_t memory = 0;
  if (os::physical_memory(memory)) {
    log_info_p(gc, init)("Memory: %zu%s",
                         byte_size_in_proper_unit(memory), proper_unit_for_byte_size(memory));
  } else {
    log_info_p(gc, init)("Memory: NA");
  }
```

stefank replied:
Alternatively:
```
  size_t memory = 0;
  (void)os::physical_memory(memory);
  log_info_p(gc, init)("Memory: %zu%s",
                       byte_size_in_proper_unit(memory), proper_unit_for_byte_size(memory));
```

toxaart replied:
Thanks, addressed.

stefank commented on src/hotspot/share/gc/z/zLargePages.cpp:
```diff
@@ -30,8 +30,11 @@ bool ZLargePages::_os_enforced_transparent_mode;
   pd_initialize();
-
-  log_info_p(gc, init)("Memory: " JULONG_FORMAT "M", os::physical_memory() / M);
+  size_t phys_mem = 0;
+  if (!os::physical_memory(phys_mem)) {
+    log_debug_p(gc, init)("os::physical_memory() failed");
+  }
+  log_info_p(gc, init)("Memory: %zuM", phys_mem / M);
```
I have the same comment about this change as I have for `GCInitLogger::print_memory`.

stefank commented on src/hotspot/share/jfr/jni/jfrJniMethod.cpp:
```diff
@@ -403,12 +403,14 @@ JVM_ENTRY_NO_ENV(jboolean, jfr_is_containerized(JNIEnv* env, jclass jvm))
 JVM_END
 
 JVM_ENTRY_NO_ENV(jlong, jfr_host_total_memory(JNIEnv* env, jclass jvm))
+  size_t phys_mem = 0;
```
This can be moved below where it is used. I would probably also be good to explicitly show that the return value is skipped.

With that said, now I've seen this pattern a number of times and think that it might be cleaner to have a function named `os::physical_memory_or_0()` (just like we have a bunch of "_or_null" functions) that returns 0 on failure. The code could then be:
```
#ifdef LINUX
  // We want the host memory, not the container limit.
  // os::physical_memory() would return the container limit.
  return os::Linux::physical_memory();
#else
  return os::physical_memory_or_0();
#endif
```
We can think about that proposal as a follow-up RFE.

toxaart replied:
I do not agree that it is clearer to have a function named   `os::physical_memory_or_0()`, because 0 does not necessarily mean a failure. It may well be a valid value for something on a system level. Not even touching various posix stuff returning 0 in case of success.

stefank commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2509,13 +2509,17 @@ WB_END
 // Physical memory of the host machine (including containers)
 WB_ENTRY(jlong, WB_HostPhysicalMemory(JNIEnv* env, jobject o))
+  size_t phys_mem = 0;
   LINUX_ONLY(return os::Linux::physical_memory();)
-  return os::physical_memory();
+  os::physical_memory(phys_mem);
+  return static_cast<jlong>(phys_mem);
```
This looks like the JFR code that doesn't want to report the container limits. We should probably have an os:: function that returns this value that both these functions wants. Maybe `os::host_physical_memory`. Something for a future RFE.

toxaart replied:
Agree.

stefank reviewed (changes requested):
I went over this and have some more comments. The two main concerns I have are:
1) I really would like to see all the UL "X failed" log lines removed from this patch.
2) Callers that discard the return value should be clearly marked that they do so. (Alternatively, provide APIs that clearly returns 0 on failure).

toxaart commented:
> 1. I really would like to see all the UL "X failed" log lines removed from this patch.

@stefank I understand your concerns. I can remove the logging for failed cases, but it will demise the proposed pattern. Without consensus on how to log such cases for JVM users, which is to be done in a later phase, I suggest to put TODO comments to the places where such logging should be done. What do you think?

toxaart commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -156,9 +156,12 @@ julong os::Bsd::available_memory() {
   if (kerr == KERN_SUCCESS) {
     // free_count is just a lowerbound, other page categories can be freed too and make memory available
     available = (vmstat.free_count + vmstat.inactive_count + vmstat.purgeable_count) * os::vm_page_size();
+  } else {
+    return false;
```
I made it to be consistent with behavior of this method implementation on other platforms. Yes, the assert just about it should be hit before that.

toxaart commented:
> Maybe we should have an `ATTRIBUTE_NODISCARD` macro with an always empty
> definition, and add it to all the functions being changed? Otherwise, once
> `[[nodiscard]]` becomes available, someone is going to have to hunt down all
> the relevant functions. Once we have C++17 we can revisit. It might take some
> preliminary call-site fixes before we can change the macro (or more likely,
> remove it and just use `[[nodiscard]]` directly).

@kimbarrett How about to change the return type from bool to some enum class? They are not implicitly convertible to integers, so should not be able to ignore the return value of a function then. And we will not need a `[[nodiscard]]` in this case.

stefank commented on src/hotspot/share/jfr/periodic/jfrPeriodic.cpp:
```diff
@@ -528,17 +528,33 @@ TRACE_REQUEST_FUNC(ThreadAllocationStatistics) {
   event.set_totalSize(totalPhysicalMemory);
-  event.set_usedSize(totalPhysicalMemory - os::available_memory());
+  size_t avail_mem = 0;
+  if (!os::available_memory(avail_mem)) {
+    log_debug(jfr, system)("os::available_memory() failed");
+  }
```
```suggestion
  // Ignore return value
  (void)os::available_memory(avail_mem);
```

stefank commented on src/hotspot/share/jfr/periodic/jfrPeriodic.cpp:
```diff
@@ -528,17 +528,33 @@ TRACE_REQUEST_FUNC(ThreadAllocationStatistics) {
-  event.set_totalSize(os::total_swap_space());
-  event.set_freeSize(os::free_swap_space());
+  size_t total_swap_space = 0;
+  if (!os::total_swap_space(total_swap_space)) {
+    log_debug(jfr, system)("os::total_swap_space() failed");
+  }
```
```suggestion
  // Ignore return value
  (void)os::total_swap_space(total_swap_space);
```

stefank commented on src/hotspot/share/jfr/periodic/jfrPeriodic.cpp:
```diff
@@ -528,17 +528,33 @@ TRACE_REQUEST_FUNC(ThreadAllocationStatistics) {
+  }
+  event.set_totalSize(static_cast<s8>(total_swap_space));
+  size_t free_swap_space = 0;
+  if (!os::free_swap_space(free_swap_space)) {
+    log_debug(jfr, system)("os::free_swap_space() failed");
+  }
```
```suggestion
  // Ignore return value
  (void)os::free_swap_space(free_swap_space);
```

stefank commented on src/hotspot/share/prims/jvmtiRedefineClasses.cpp:
```diff
@@ -1358,10 +1358,13 @@ jvmtiError VM_RedefineClasses::load_new_class_versions() {
     InstanceKlass* the_class = get_ik(_class_defs[i].klass);
-
+    size_t avail_mem = 0;
+    if (!os::available_memory(avail_mem)) {
+      log_debug(redefine, class, load)("os::available_memory() failed");
+    }
```
```suggestion
    // Ignore return value
    (void)os::available_memory(avail_mem);
```

stefank commented on src/hotspot/share/prims/jvmtiRedefineClasses.cpp:
```diff
@@ -4435,9 +4440,13 @@ void VM_RedefineClasses::redefine_single_class(Thread* current, jclass the_jclas
     // increment the classRedefinedCount field in the_class and in any
     // direct and indirect subclasses of the_class
+    size_t avail_mem = 0;
+    if (!os::available_memory(avail_mem)) {
+      log_debug(redefine, class, load)("os::available_memory() failed");
+    }
```
```suggestion
    // Ignore return value
    (void)os::available_memory(avail_mem);
```

toxaart replied:
Removed.

stefank commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -1520,13 +1520,20 @@ void Arguments::set_heap_size() {
   if (override_coop_limit) {
     if (FLAG_IS_DEFAULT(MaxRAM)) {
-      phys_mem = os::physical_memory();
+      if (!os::physical_memory(physical_mem_val)) {
+        log_debug(os)("os::physical_memory() failed");
+      }
```
```suggestion
      // Ignore return value
      (void)os::physical_memory(physical_mem_val);
```

toxaart replied:
Removed.

stefank commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -1520,13 +1520,20 @@ void Arguments::set_heap_size() {
     }
   } else {
-    phys_mem = FLAG_IS_DEFAULT(MaxRAM) ? MIN2(os::physical_memory(), (julong)MaxRAM)
+    if (!os::physical_memory(physical_mem_val)) {
+      log_debug(os)("os::physical_memory() failed");
+    }
```
```suggestion
    // Ignore return value
    (void)os::physical_memory(physical_mem_val);
```

toxaart replied:
Removed.

stefank commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1183,9 +1183,13 @@ void os::print_summary_info(outputStream* st, char* buf, size_t buflen) {
   st->print("%s, ", buf);
-  size_t mem = physical_memory()/G;
+  size_t phys_mem = 0;
+  if (!physical_memory(phys_mem)) {
+    log_debug(os)("os::physical_memory() failed");
+  }
```
```suggestion
  // Ignore return value
  (void)physical_memory(phys_mem);
```

toxaart replied:
Removed.

stefank commented on src/hotspot/share/runtime/os.cpp:
```diff
@@ -1940,10 +1944,13 @@ bool os::is_server_class_machine() {
   const julong missing_memory   = 256UL * M;
-
+  size_t phys_mem = 0;
+  if (!os::physical_memory(phys_mem)) {
+    log_debug(os)("os::physical_memory() failed");
+  }
```
```suggestion
  // Ignore return value
  (void)os::physical_memory(phys_mem);
```

stefank commented on src/hotspot/share/services/heapDumper.cpp:
```diff
@@ -2612,7 +2612,11 @@ int HeapDumper::dump(const char* path, outputStream* out, int compression, bool
     // Lets ensure we have at least 20MB per thread.
-    julong max_threads = os::free_memory() / (20 * M);
+    size_t free_memory = 0;
+    if (!os::free_memory(free_memory)) {
+      log_debug(heapdump)("os::free_memory() failed");
+    }
```
```suggestion
    (void)os::free_memory(free_memory);
```

stefank reviewed (commented):
I proposed the usage of (void) but you instead used static_cast<void>. We have an overwhelmingly larger set of (void) usages so I think you should stick to using the prevalent form.

toxaart commented:
> I proposed the usage of (void) but you instead used static_cast. We have an overwhelmingly larger set of (void) usages so I think you should stick to using the prevalent form.

I think I addressed all points. As for (void) vs `static_cast<void>` conversion, the only benefit I see is shortness. But okay.

stefank commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -1520,13 +1520,15 @@ void Arguments::set_heap_size() {
   if (override_coop_limit) {
     if (FLAG_IS_DEFAULT(MaxRAM)) {
-      phys_mem = os::physical_memory();
+      (void)os::physical_memory(physical_mem_val);
```
This is one of the places where things seems to break down if we ever get a failure from os::physical_memory(). Before this patch I guess you would get random values and after this patch you will propagate 0 through this function. So, I guess it is not worse than before, but if we really can get a failure here then we do have an existing bug here.

With that said, when looking at this I'm wondering if there is any scenario where we can hit an error when calling the os::physical_memory? I understand that the OS APIs says that we can hit an error, but I wonder if that's only a theoretical issue? If it turns out that the APIs in effect never will fail then we can verify that they succeed during the JVM initialization and then remove the error code from os os::physical_memory() and let it only return a size_t instead. If we do this we can clean away a lot of the error handling code that is added in this PR.

The places that I see that could theoretically fail is:

1) On AIX, but there's an `assert(0` in there to show that this really doesn't happen.

2) On Windows if GlobalMemoryStatusEx(&ms) fails. The question is will this call fail for any reason on Windows? Maybe someone from Microsoft could help and tell if there ever is a reason why this code would fail:
```
  MEMORYSTATUSEX ms;
  ms.dwLength = sizeof(ms);
  GlobalMemoryStatusEx(&ms);
```

toxaart replied:
1) As for sanity checks:

- In `os::init()` in os_linux.cpp there is a sanity check for sysconf. It is done just before calling` Linux::initialize_system_info()` where `_physical_memory` is set. 

- In `os::Aix::initialize_system_info()` there is already a sanity check.

- In `os::Bsd::initialize_system_info()` there is a check for success of `sysctl`, but a default value 256 * 1024 * 1024 is assigned to `_physical_memory`. I do not know why it is done this way. Maybe this needs to be addressed separately.

- In `os::win32::initialize_system_info()` there is a call to `GlobalMemoryStatusEx`, but above we've came to conclusion that it never fails. 

So all OS sysinfo init routines have a sort of sanity check, except BSD, but I guess it was done intentionally. 

2) Addressed.

3) As for the rest of the methods:

- `available_memory()` can fail on Linux only if sysinfo fails, which can happen. Since it can fail on one OS, then the function should return both success/failure and the resulting value.
- `used_memory()` is never used in the code. Maybe it should be completely removed? 
- `free_memory()`: the same argument applies as to  `available_memory()`.
- `total_swap_space()`: the same argument applies as to  `available_memory()`.
- `free_swap_space()`: the same argument applies as to  `available_memory()`.

stefank reviewed (changes requested):
From my POV I think this starts to look OK except for a few nits.

I have a more lengthy comment below in Arguments::set_heap_size where I'm wondering if we really need to be returning an error in os::pysical_memory(). The question is mostly directed to Microsoft devs who might have insight into the memory API we use to query to get amount physical memory.

We could treat that question as a follow-up question and then maybe clean out the added error handling code (and places ignoring the error). Or we could try to resolve this before this PR is pushed so that we don't temporarily add all this extra error handling/ignoring code. What do you (and other reviewers) think?

toxaart commented on src/hotspot/share/gc/shared/gcInitLogger.cpp:
```diff
@@ -62,9 +62,12 @@ void GCInitLogger::print_cpu() {
+    log_info_p(gc, init)("Memory: %zu%s", byte_size_in_proper_unit(phys_mem), proper_unit_for_byte_size(phys_mem));
+  } else {
+    log_info_p(gc, init)("Memory: NA");
+  }
```
Addressed.

toxaart commented on src/hotspot/share/gc/z/zLargePages.cpp:
```diff
@@ -30,8 +30,12 @@ bool ZLargePages::_os_enforced_transparent_mode;
+    log_info_p(gc, init)("Memory: %zu%s", byte_size_in_proper_unit(phys_mem), proper_unit_for_byte_size(phys_mem));
+  } else {
+    log_info_p(gc, init)("Memory: NA");
+  }
```
Addressed.

toxaart commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -1507,7 +1507,7 @@ static const size_t DefaultHeapBaseMinAddress = HeapBaseMinAddress;
 void Arguments::set_heap_size() {
   julong phys_mem;
-
+  size_t physical_mem_val = 0;
```
Addressed.

dholmes-ora commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -1504,6 +1504,7 @@ static const size_t DefaultHeapBaseMinAddress = HeapBaseMinAddress;
 
 void Arguments::set_heap_size() {
   julong phys_mem;
+  size_t physical_mem_val = 0;
```
I'm not seeing why you needed to introduce this as a temporary. ??

toxaart replied:
This is a leftover from the latest change `os::physical_memory(size_t& val) -> size_t os::physical_memory()`. Addressed in the latest commit.

dholmes-ora commented on src/hotspot/share/services/heapDumper.cpp:
```diff
@@ -2612,7 +2612,9 @@ int HeapDumper::dump(const char* path, outputStream* out, int compression, bool
     // For the OOM handling we may already be limited in memory.
     // Lets ensure we have at least 20MB per thread.
-    julong max_threads = os::free_memory() / (20 * M);
+    size_t free_memory = 0;
+    (void)os::free_memory(free_memory);
```
```suggestion
    size_t free_memory = 0;
    // Return value ignored - defaulting to 0 on failure.
    (void)os::free_memory(free_memory);
```

toxaart replied:
Addressed.

dholmes-ora commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -1360,4 +1360,9 @@ std::add_rvalue_reference_t<T> declval() noexcept;
 
+//----------------------------------------------------------------------------------------------------
+
+// Dummy placeholder for use of [[nodiscard]]
+#define ATTRIBUTE_NODISCARD
```
Stay-tuned we may actually be able to use `[[nodiscard]]` for real.

dholmes-ora reviewed (approved):
Overall I'm okay with the way the API has turned out here. Naturally it would be better if no errors were possible, but I think containers prevent that from being the case.

I have a lot of stylistic comments, many pre-existing, and some notes for future cleanups. But happy to hit approve as a vote of confidence.

Thanks

toxaart commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2511,13 +2511,17 @@ WB_END
-  return os::physical_memory();
+  LINUX_ONLY(return static_cast<jlong>(os::Linux::physical_memory());)
+  size_t phys_mem = os::physical_memory();
+  return static_cast<jlong>(phys_mem);
```
Addressed.

toxaart commented on src/hotspot/share/prims/jvmtiRedefineClasses.cpp:
```diff
@@ -4435,9 +4436,11 @@ void VM_RedefineClasses::redefine_single_class(Thread* current, jclass the_jclas
     // increment the classRedefinedCount field in the_class and in any
     // direct and indirect subclasses of the_class
+    size_t avail_mem = 0;
+    (void)os::available_memory(avail_mem);
```
Addressed.

toxaart commented on src/hotspot/share/prims/jvmtiRedefineClasses.cpp:
```diff
@@ -1525,9 +1526,9 @@ jvmtiError VM_RedefineClasses::load_new_class_versions() {
       }
     }
-
+    (void)os::available_memory(avail_mem);
```
Addressed.

toxaart commented on src/hotspot/share/prims/jvmtiRedefineClasses.cpp:
```diff
@@ -1358,10 +1358,11 @@ jvmtiError VM_RedefineClasses::load_new_class_versions() {
     InstanceKlass* the_class = get_ik(_class_defs[i].klass);
-
+    size_t avail_mem = 0;
+    (void)os::available_memory(avail_mem);
```
Addressed.

toxaart commented on src/hotspot/share/jfr/periodic/jfrPeriodic.cpp:
```diff
@@ -528,17 +528,24 @@ TRACE_REQUEST_FUNC(ThreadAllocationStatistics) {
+  (void)os::total_swap_space(total_swap_space);
+  event.set_totalSize(static_cast<s8>(total_swap_space));
+  size_t free_swap_space = 0;
+  (void)os::free_swap_space(free_swap_space);
```
Addressed.

toxaart commented on src/hotspot/share/jfr/periodic/jfrPeriodic.cpp:
```diff
@@ -528,17 +528,24 @@ TRACE_REQUEST_FUNC(ThreadAllocationStatistics) {
   event.set_totalSize(totalPhysicalMemory);
-  event.set_usedSize(totalPhysicalMemory - os::available_memory());
+  size_t avail_mem = 0;
+  (void)os::available_memory(avail_mem);
```
Addressed.

toxaart commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -255,97 +256,119 @@ julong os::Linux::available_memory() {
   // less than host_free_swap(). The upper bound of free swap needs to be the lower of the two.
-  jlong host_free_swap_val = MIN2(os::total_swap_space(), host_free_swap());
-  assert(host_free_swap_val >= 0, "sysinfo failed?");
+  size_t total_swap_space = 0;
+  bool total_swap_space_ok = os::total_swap_space(total_swap_space);
+  size_t host_free_swap = 0;
+  bool host_free_swap_ok = host_free_swap_f(host_free_swap);
+  if (!total_swap_space_ok || !host_free_swap_ok) {
+    assert(false, "sysinfo failed ? ");
+    return false;
+  }
```
I suggest to keep results as locals, see my previous comment.

toxaart replied:
With the locals it would be visible in debugging which of those methods failed if any. 

Addressed as suggested.

toxaart commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -255,97 +256,119 @@ julong os::Linux::available_memory() {
+  if (OSContainer::is_containerized() && OSContainer::memory_and_swap_limit_in_bytes() > 0) {
     if (OSContainer::memory_limit_in_bytes() > 0) {
-      return (jlong)(OSContainer::memory_and_swap_limit_in_bytes() - OSContainer::memory_limit_in_bytes());
+      value = static_cast<size_t>(OSContainer::memory_and_swap_limit_in_bytes() - OSContainer::memory_limit_in_bytes());
```
Addressed.

toxaart commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -862,35 +862,41 @@ bool os::win32::available_memory(size_t& value) {
     return true;
   } else {
-    return false;
+    errno = ::GetLastError();
```
You are correct, there is no relationship between `errno` and `GetLastError()`.  Addressed.

I created a separate issue to fix this inconsistency in other places.

dholmes-ora replied:
I had forgotten that this oddity was also raised in the review of [JDK-8350869](https://bugs.openjdk.org/browse/JDK-8350869). If it is part of the `os::xxx` function's specification to set `errno` on failure then we would need to do something to achieve that (which I think was the intended case with something like `os::stat`) though as previously noted the set of allowed values for `errno` and `GetLastErrror` are not the same. But the API's in question here are not specified that way.

toxaart commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -294,7 +294,9 @@ bool os::Linux::free_memory(size_t& value) {
-      value = static_cast<size_t>(OSContainer::memory_and_swap_limit_in_bytes() - OSContainer::memory_limit_in_bytes());
+      jlong memory_and_swap_limit_in_bytes = OSContainer::memory_and_swap_limit_in_bytes();
+      jlong memory_limit_in_bytes = OSContainer::memory_limit_in_bytes();
+      value = static_cast<size_t>(memory_and_swap_limit_in_bytes - memory_limit_in_bytes);
```
Right, I changed the way it is done as suggested.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -255,97 +256,119 @@ julong os::Linux::available_memory() {
+  size_t host_free_swap = 0;
+  bool host_free_swap_ok = host_free_swap_f(host_free_swap);
+  if (!total_swap_space_ok || !host_free_swap_ok) {
+    assert(false, "sysinfo failed ? ");
```
If we are going to assert on failure then we should report as much information about the failure as possible.

toxaart commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -1360,4 +1360,9 @@ std::add_rvalue_reference_t<T> declval() noexcept;
+//----------------------------------------------------------------------------------------------------
+
+// Dummy placeholder for use of [[nodiscard]]
+#define ATTRIBUTE_NODISCARD
```
Okay, moved it up.

toxaart commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -848,39 +848,59 @@ jlong os::elapsed_frequency() {
+    return true;
+  } else {
+    DWORD errcode  = ::GetLastError();
+    log_debug(os)("total_swap_space() failed to GlobalMemoryStatusEx: GetLastError->%lu.", errcode);
```
Addressed.

toxaart commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -848,39 +848,59 @@ jlong os::elapsed_frequency() {
+    return true;
+  } else {
+    DWORD errcode  = ::GetLastError();
+    log_debug(os)("free_swap_space() failed to GlobalMemoryStatusEx: GetLastError->%lu.", errcode);
```
Addressed.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -292,15 +292,18 @@ bool os::Linux::free_memory(size_t& value) {
   struct sysinfo si;
   int ret = sysinfo(&si);
   if (ret != 0) {
+    assert(false, "sysinfo failed in total_swap_space()?");
```
We should report the value of errno in these cases.
```suggestion
    assert(false, "sysinfo failed in total_swap_space(): %s", os::strerror(errno));
```

toxaart replied:
Added `errno `reporting as suggested.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -862,35 +862,38 @@ bool os::win32::available_memory(size_t& value) {
     return true;
   } else {
-    return false;
+    assert(false, "GlobalMemoryStatusEx failed in win32::available_memory().");
```
These should report the value of `::GetLastError()`

toxaart replied:
Added error reporting as suggested.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -862,35 +862,38 @@ bool os::win32::available_memory(size_t& value) {
   } else {
-    return false;
+    assert(false, "GlobalMemoryStatusEx failed in win32::available_memory().");
+    return res == TRUE;
```
```suggestion
    return false;
```
Please restore all these to "return false" as we know that `res == false` here.

toxaart replied:
Addressed.

dholmes-ora reviewed (changes requested):
Asserts should always provide as much error detail as is available - otherwise when they fail we do not know why.

Thanks

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -866,8 +866,8 @@ bool os::win32::available_memory(size_t& value) {
   } else {
-    assert(false, "GlobalMemoryStatusEx failed in win32::available_memory().");
-    return res == TRUE;
+    assert(false, "GlobalMemoryStatusEx failed in win32::available_memory(): %u", ::GetLastError());
```
```suggestion
    assert(false, "GlobalMemoryStatusEx failed in win32::available_memory(): %lu", ::GetLastError());
```
There is no consistency here but the primary ways we print `GetLastError` are using `%lu`, `%ld` or `%d`.

toxaart replied:
Addressed by usage of `%lu`.

toxaart commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -1360,4 +1363,7 @@ std::add_rvalue_reference_t<T> declval() noexcept;
 bool IEEE_subnormal_handling_OK();
 
+//----------------------------------------------------------------------------------------------------
+
```
I removed all added lines, it is a leftover from the initial placing of the `ATTRIBUTE_NODISCARD` dummy placeholder.

toxaart commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -4145,8 +4162,12 @@ void os::win32::initialize_system_info() {
-  _physical_memory = ms.ullTotalPhys;
+  BOOL res = GlobalMemoryStatusEx(&ms);
+  if (res != TRUE)
+  {
```
Thanks for spotting, addressed.

stefank commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -848,39 +848,56 @@ jlong os::elapsed_frequency() {
+    value = static_cast<size_t>(ms.ullTotalPageFile);
+    return true;
+  } else {
+    assert(false, "GlobalMemoryStatusEx failed in win32::total_swap_space(): %lu", ::GetLastError());
```
```suggestion
    assert(false, "GlobalMemoryStatusEx failed in os::total_swap_space(): %lu", ::GetLastError());
```

stefank commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -848,39 +848,56 @@ jlong os::elapsed_frequency() {
+    value = static_cast<size_t>(ms.ullAvailPageFile);
+    return true;
+  } else {
+    assert(false, "GlobalMemoryStatusEx failed in win32::free_swap_space(): %lu", ::GetLastError());
```
```suggestion
    assert(false, "GlobalMemoryStatusEx failed in os::free_swap_space(): %lu", ::GetLastError());
```

stefank commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -848,39 +848,56 @@ jlong os::elapsed_frequency() {
+    value = static_cast<size_t>(ms.ullAvailPhys);
+    return true;
+  } else {
+    assert(false, "GlobalMemoryStatusEx failed in os::available_memory(): %lu", ::GetLastError());
```
Yes and git will remind you about that if you try to push without doing so.

dholmes-ora commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -56,6 +56,9 @@ class oopDesc;
 #define ATTRIBUTE_FLATTEN
 #endif
 
+// Dummy placeholder for use of [[nodiscard]]
+#define ATTRIBUTE_NODISCARD
```
I think you can just use `[[nodiscard]]` directly now. We are going to allow it and the guide is just going through the official update process.

toxaart replied:
Removed as suggested.

dholmes-ora reviewed (approved):
One comment below but otherwise I think everything I have been concerned about has been addressed. Thanks for your patience.

stefank reviewed (approved):
Looks good (if this still compiles with the `[[nodiscard]]` changes).

toxaart commented:
I have re-run tests in tiers 1-3 on all available platforms, no problems detected after the latest changes on this branch.

dholmes-ora commented:
/sponsor
